File: services/auth-service/pkg/response/response.go
// Package response provides canonical JSON response builders.
// ALL handlers MUST use these helpers — never write gin.H directly.
//
// Rules (enforced by code review):
//  1. Success responses:  response.OK / response.Created
//  2. Error responses:    response.Error (pass an *apperrors.AppError)
//  3. NEVER expose raw DB errors, driver messages, or SQL to the client.
//     Wrap DB errors in apperrors.ErrInternal before they leave the service layer.
package response

import (
	"net/http"

	"github.com/gin-gonic/gin"

	apperrors "github.com/english-web/auth-service/pkg/errors"
)

// SuccessEnvelope is the canonical success response shape.
//
//	{ "success": true, "data": <T>, "message": "<optional>" }
type SuccessEnvelope struct {
	Success bool   `json:"success"`
	Data    any    `json:"data"`
	Message string `json:"message,omitempty"`
}

// ErrorEnvelope is the canonical error response shape.
//
//	{ "success": false, "error": { "code": "...", "message": "..." } }
type ErrorEnvelope struct {
	Success bool         `json:"success"`
	Error   errorPayload `json:"error"`
}

type errorPayload struct {
	Code    string `json:"code"`
	Message string `json:"message"` // Vietnamese, user-facing
}

// OK writes HTTP 200 with data.
func OK(c *gin.Context, data any) {
	c.JSON(http.StatusOK, SuccessEnvelope{Success: true, Data: data})
}

// OKWithMessage writes HTTP 200 with data and a message.
func OKWithMessage(c *gin.Context, data any, message string) {
	c.JSON(http.StatusOK, SuccessEnvelope{Success: true, Data: data, Message: message})
}

// Created writes HTTP 201 with data.
func Created(c *gin.Context, data any) {
	c.JSON(http.StatusCreated, SuccessEnvelope{Success: true, Data: data})
}

// NoContent writes HTTP 204 with no body.
func NoContent(c *gin.Context) {
	c.Status(http.StatusNoContent)
}

// Error writes the appropriate HTTP status + error envelope.
// Only accepts *apperrors.AppError to prevent leaking raw errors.
func Error(c *gin.Context, err *apperrors.AppError) {
	c.JSON(err.StatusCode, ErrorEnvelope{
		Success: false,
		Error: errorPayload{
			Code:    err.Code,
			Message: err.Message,
		},
	})
	c.Abort()
}

// InternalError writes HTTP 500, logging the raw error server-side only.
// Never expose the raw error message to the client.
func InternalError(c *gin.Context) {
	Error(c, apperrors.ErrInternal)
}

File: services/ai-service/app/core/response.py
"""
Canonical HTTP response helpers.

ALL routers MUST use these helpers.
Do NOT construct response dicts manually in routers.

Success:  response.ok(data) / response.created(data)
Error:    response.error(app_error)         <- only AppError allowed
Internal: response.internal()              <- logs raw error, returns generic message
"""

from __future__ import annotations

import logging
from typing import Any

from app.core.errors import ERR_INTERNAL, AppError
from fastapi import Response
from fastapi.responses import JSONResponse

logger = logging.getLogger(__name__)


def ok(data: Any, message: str = "") -> JSONResponse:
    body: dict[str, Any] = {"success": True, "data": data}
    if message:
        body["message"] = message
    return JSONResponse(content=body, status_code=200)


def created(data: Any) -> JSONResponse:
    return JSONResponse(content={"success": True, "data": data}, status_code=201)


def no_content() -> Response:
    return Response(status_code=204)


def error(app_error: AppError) -> JSONResponse:
    return JSONResponse(
        content={
            "success": False,
            "error": {"code": app_error.code, "message": app_error.message},
        },
        status_code=app_error.status_code,
    )


def internal(raw_error: Exception | None = None) -> JSONResponse:
    """Log the raw error server-side, return generic 500 to client."""
    if raw_error:
        logger.exception("Unhandled internal error", exc_info=raw_error)
    return error(ERR_INTERNAL)
